### PR TITLE
Prevent some race conditions

### PR DESCRIPTION
**Pull request by schustmi**
## Describe changes
This PR adds some unique constraint to avoid race conditions when creating entities, and also moves the retry logic for model version creation to the server.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/exceptions.py [hunk 1]
```diff
@@ -168,6 +168,10 @@ class EntityExistsError(ZenMLBaseException):
     """Raised when trying to register an entity that already exists."""
 
 
+class EntityCreationError(ZenMLBaseException):
+    """Raised when failing to create an entity."""
+
+
```
Comment (avishniakov) [inline src/zenml/exceptions.py:174]: I bet we decided to stop creating custom exceptions. At least @stefannica always told me some for the core.

**Review by avishniakov — APPROVED** [review]
Only a minor comment, assuming tests will pass smoothly.

File: src/zenml/exceptions.py (hunk 1 above)
Comment (schustmi) [reply src/zenml/exceptions.py]: I talked to @stefannica and we agreed it makes sense in this case.

Comment (avishniakov) [reply src/zenml/exceptions.py]: ok

**Pull request merged**